### PR TITLE
Dev era5 variables

### PR DESCRIPTION
**Pull request by robert-DL**
This is a rather large patch. 

I added the option to load more variables from the ERA5 dataset. For now I limited my attention to the variables available in the 1.5° Dataset. The normalization statistics are now loaded from json files. The contain the normalization data from graphcast. These values are slightly different to the ones from pangu, as pangu used more years to compute the statistics. 

I had to change dataloaders, test scripts, metrics and the logging behaviour for variables. 

As people might want to use sea variables, I included a nan_to_zero mechanism. 
The diffusion module is untouched for now and a similar patch is needed in the case of diffusion. 

The normalization values are now calculated in a separate module, which is instantiated with hydra in Era5Forecast and in ForecastingModule. 

As I am not sure about all changes and there validity, I am open to suggestions and improvements here.

### CODE REVIEW

File: geoarches/metrics/metric_base.py [hunk 1]
```diff
@@ -70,7 +70,7 @@ def wmse(self, x: torch.Tensor, y: torch.Tensor | int = 0):
             y: targets with shape (..., lat, lon)
         """
         lat_coeffs = self.compute_lat_weights_fn(latitude_resolution=x.shape[-2]).to(x.device)
-        return (x - y).pow(2).mul(lat_coeffs).mean((-2, -1))
+        return (x - y).pow(2).mul(lat_coeffs).nanmean((-2, -1))
```
Comment (14renus) [inline geoarches/metrics/metric_base.py:73]: I still have a slight problem with just ignoring nans.

maybe, can we do a check that x is not nan anywhere were y is not nan?

Comment (robert-DL) [reply geoarches/metrics/metric_base.py:73]: I can implement that.

File: geoarches/configs/stats/graphcast.yaml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+module:
+  _target_: geoarches.utils.norm_utils.NormModule
+  variables:
```
Comment (14renus) [inline geoarches/configs/stats/graphcast.yaml:3]: Nice, I like the idea of a separate class to handle normalization.

However, feels like variables, levels should live in the dataloader config directly and then passed into NormModule, since these are used generally in dataloading (beyond just normalization) 

and we can pass in `norm_scheme` as the sole argument to the dataloader and the dataloader takes care of instantiating NormModule with the variables, levels, and norm_scheme.

And I think loss should be separated from normalization logic, as these two things are not related.

File: geoarches/configs/stats/graphcast.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        - specific_humidity
+        - vertical_velocity
+  loss_weight_per_variable:
+    surface: [0.1, 0.1, 1.0, 0.1]
+    level: [1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
+    
+  levels: [50, 100, 150, 200, 250, 300, 400, 500, 600, 700, 850, 925, 1000]
+  norm_scheme: graphcast
+
+loss_coeffs:
```
Comment (14renus) [inline geoarches/configs/stats/graphcast.yaml:23]: compute_loss_coeffs_args:

is clearer what these are used for

File: geoarches/lightning_modules/forecast.py [hunk 4]
```diff
@@ -165,8 +126,13 @@ def loss(self, pred, gt, multistep=False, **kwargs):
 
             loss_coeffs.apply(lambda x: x * future_coeffs)
 
-        weighted_error = (pred - gt).abs().pow(self.pow).mul(loss_coeffs)
+        
 
+        pred = TensorDict({k: (~torch.isnan(gt[k])).float() * v for k, v in pred.items()}, batch_size=gt.batch_size)
```
Comment (14renus) [inline geoarches/lightning_modules/forecast.py:131]: consider adding a function to tensordict_utils.py instead and calling it here twice (ie. tdict = zero_out_nans(tdict))

`batch_size=gt.batch_size` and `batch_size=pred.batch_size` are switched?

File: poetry.lock
Comment (14renus) [inline poetry.lock:1]: is this change needed? 
if so this would need to be added to pyproject.toml first

File: tests/dataloaders/test_era5.py
Comment (14renus) [inline tests/dataloaders/test_era5.py:1]: can we add an additional test here that normalization for both pangu and gencast works?

mainly loading an example produces expected shape with no errors
feel free to copy tests here

File: geoarches/dataloaders/era5.py [hunk 5]
```diff
@@ -336,25 +412,33 @@ def __getitem__(self, i, normalize=True):
...
+            )
 
         # Load multiple future timestamps if specified.
         if self.multistep > 1:
             future_states = []
             for k in range(1, self.multistep + 1):
-                future_states.append(super().__getitem__(i + k * T // self.timedelta))
+                future_states.append(
+                    super().__getitem__(
+                        i + k * T // self.timedelta, interpolate_nans=False, warning_on_nan=False
```
[the remark this answers is not in the document]
Comment (robert-DL) [reply geoarches/dataloaders/era5.py:433]: The problem is that we want the input states to be interpolated but not the target states. The reason is that we need the information about NaN positions in the training to zero them out before calculating the loss. Therefore I suggest to keep the settings for future states and next state accordingly.

File: geoarches/configs/stats/graphcast.yaml (hunk 2 above)
Comment (robert-DL) [reply geoarches/configs/stats/graphcast.yaml:3]: That is true. I did it this way since I included the loss normalization in the module and therefore, NormModule is instantiated in ForecastModule too. The information therefore lives in the normalization module. However, I can make variables and levels part of the dataloader config and interpolate within the stats yaml file instead.

File: geoarches/dataloaders/netcdf.py [hunk 6]
```diff
@@ -142,18 +147,21 @@ def __getitem__(self, i, return_timestamp=False, interpolate_nans=False, warning
...
-
-        if interpolate_nans is None:
-            interpolate_nans = self.interpolate_nans
-
         if interpolate_nans:
-            obsi = obsi.fillna(value=obsi.mean(dim=["latitude", "longitude"], skipna=True))
+            # check if either latitude or longitude dimensions exist
+            # or lat and lon are dimension names
+            if "latitude" in obsi.dims and "longitude" in obsi.dims:
+                obsi = obsi.fillna(value=obsi.mean(dim=["latitude", "longitude"], skipna=True))
```
Comment (14renus) [inline geoarches/dataloaders/netcdf.py:154]: i think this should be a change in the config/arg for the dataloader like:
latitude_dim_name='latitude'
longitude_dim_name='longitude'

and used here and also in `convert_to_tensordict` and `'convert_to_xarray'

File: tests/dataloaders/test_era5_graphcast.py
Comment (14renus) [inline tests/dataloaders/test_era5_graphcast.py:1]: I'm going to be annoying and ask for a file comment just explaining this is testing era5 dataset with graphcast normalization. 
(this is for my forgetful brain too)

**Review by 14renus — CHANGES_REQUESTED** [review]
thank you this is great!!!

File: geoarches/dataloaders/netcdf.py [hunk 7]
```diff
@@ -31,11 +32,16 @@ def __init__(
...
         variables: Dict[str, List[str]],
-        dimension_indexers: Dict[str, list] | None = None,
+        dimension_indexers: Dict[str, list] | None,
         filename_filter: Callable = lambda _: True,  # condition to keep file in dataset
         return_timestamp: bool = False,
-        warning_on_nan: bool = False,
+        warning_on_nan: bool = True,
         limit_examples: int | None = None,
+        interpolate_nans: bool = True,
+        latitude_dim_name: str = "None",
```
[the remark this answers is not in the document]
Comment (robert-DL) [reply geoarches/dataloaders/netcdf.py:41]: Was a mistake by me. I also used default values for Era5Dataset and Era5Forecast but missed it here.

File: geoarches/dataloaders/era5.py [hunk 8]
```diff
@@ -204,6 +208,10 @@ def __init__(
             return_timestamp=return_timestamp,
             warning_on_nan=warning_on_nan,
             interpolate_nans=interpolate_nans,
+            latitude_dim_name=latitude_dim_name,
+            longitude_dim_name=longitude_dim_name,
+            level_dim_name=level_dim_name,
+            time_dim_name=time_dim_name,
         )
 
     def convert_to_tensordict(self, xr_dataset):
```
[the remark this answers is not in the document]
Comment (robert-DL) [reply geoarches/dataloaders/era5.py:217]: True, oversaw that :)

File: geoarches/dataloaders/era5.py [hunk 9]
```diff
@@ -118,15 +193,25 @@ def __init__(
...
-            warning_on_nan=True,
+            warning_on_nan=warning_on_nan,
+            interpolate_nans=interpolate_nans,
+            latitude_dim_name=latitude_dim_name,
+            longitude_dim_name=longitude_dim_name,
+            level_dim_name=level_dim_name,
+            time_dim_name=time_dim_name,
         )
 
     def convert_to_tensordict(self, xr_dataset):
```
[the remark this answers is not in the document]
Comment (14renus) [reply geoarches/dataloaders/era5.py:217]: yes, that would work well!

File: geoarches/dataloaders/netcdf.py [hunk 10]
```diff
@@ -133,16 +142,23 @@ def convert_to_tensordict(self, xr_dataset):
         """
         # Optionally select dimensions.
         if self.dimension_indexers and not self.already_ran_index_selection:
-            xr_dataset = xr_dataset.sel(self.dimension_indexers)
+            indexers = {
```
Comment (robert-DL) [inline geoarches/dataloaders/netcdf.py:145]: The thing here is that it is not possible to select single coordinates as sel uses slice here, so another fix is needed. 
Maybe one  introduces a method that properly does this stuff and decides, depedning on what is given in dimension indexers if slice is used or single coordinates are selected.

Comment (dataymeric) [reply geoarches/dataloaders/netcdf.py:145]: so for example `.sel(level=slice(50, 50))` will select only one coordinate, so we might be able to exploit this as a workaround?

Comment (dataymeric) [reply geoarches/dataloaders/netcdf.py:145]: right, I did not consider this
I guess the "easiest" way would be something along the line:

```python
indexers = {}
for dim_alias, (xr_dim, value) in self.dimension_indexers.items():
    if dim_alias == 'time':
        continue
    if value is None:
        indexers[xr_dim] = slice(None)
    elif isinstance(value, tuple):
        indexers[xr_dim] = slice(*value)
    else:
        indexers[xr_dim] = value  # can be scalar or list
```

**Review by 14renus — COMMENTED** [review]
Hey Robert, thanks for the patience and all the changes. I think we are converging now (:

My only remaining comment is about `dimension_indexers`.
We should not use slice as the default indexing method for `dimension_indexers`. Users should only have to specify `slice` if they want to use it. This allows the user to also just give a list.

File: geoarches/dataloaders/netcdf.py [hunk 11]
```diff
@@ -32,24 +38,32 @@ def __init__(
         self,
         path: str,
         variables: Dict[str, List[str]],
-        dimension_indexers: Dict[str, list] | None,
+        dimension_indexers: Dict[str, list] = None,
```
[the remark this answers is not in the document]
Comment (robert-DL) [reply geoarches/dataloaders/netcdf.py:41]: I don't remember why but there was a reason why I did not used default_dimension_indexers as the argument.

File: geoarches/dataloaders/netcdf.py (hunk 10 above)
Comment (14renus) [reply geoarches/dataloaders/netcdf.py:145]: Yeah, I think for now since we will only train on one resolution, we can just specify in code for now until we have a working solution for the dimension_indexers in hydra config. otherwise this PR might never end (:

**Review by 14renus — DISMISSED** [review]

File: geoarches/evaluation/eval_multistep.py [hunk 12]
```diff
@@ -151,7 +152,10 @@ def main():
             "Need to provide surface and/or level variables to load using --surface_vars and --level_vars."
         )
 
+    
     # Groundtruth.
+    dimension_indexers = {k: v for k, v in default_dimension_indexers.items()}
+    dimension_indexers['level'] = ('level', [500, 700, 850])  # Use only these pressure levels.
```
[the remark this answers is not in the document]
Comment (robert-DL) [reply geoarches/evaluation/eval_multistep.py:158]: I could interpolate the dimension indexers from the stats file or vice versa and put into dimension indexers. But if you once start specifying dim_indexers, they have to be part of the repo completely

File: geoarches/evaluation/eval_multistep.py [hunk 13]
```diff
@@ -151,7 +152,10 @@ def main():
             "Need to provide surface and/or level variables to load using --surface_vars and --level_vars."
         )
 
+    
     # Groundtruth.
+    dimension_indexers = {k: v for k, v in default_dimension_indexers.items()}
```
Comment (14renus) [inline geoarches/evaluation/eval_multistep.py:157]: if you need to copy, this is better:

`dimension_indexers = default_dimension_indexers.copy()`

**Review by 14renus — APPROVED** [review]

File: geoarches/dataloaders/netcdf.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (14renus) [reply geoarches/dataloaders/netcdf.py:42]: TODO: will follow up in next PR

File: geoarches/configs/stats/graphcast.yaml (hunk 2 above)
Comment (14renus) [reply geoarches/configs/stats/graphcast.yaml:3]: TODO if wanted - Needed for both dataloader and loss stats, doesn't matter too much where these are set.

File: geoarches/dataloaders/era5.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (14renus) [reply geoarches/dataloaders/era5.py:460]: TODO: follow up in next PR
